casper/abstract_view.py

```diff
@@ -76,13 +85,32 @@ def resolve_waiting_messages(self, message):
 
             if not any(self.missing_message_dependencies[dependent_hash]):
                 new_message = self.pending_messages[dependent_hash]
-                self.add_to_justified_messages(new_message)
                 self.resolve_waiting_messages(new_message)
-                del self.missing_message_dependencies[new_message.hash]
-                del self.pending_messages[new_message.hash]
+
+                self.mark_message_as_fully_received(new_message)
 
         del self.dependents_of_message[message.hash]
 
+    def mark_message_as_fully_received(self, message):
```

Comment:
renamed from `add_to_justified_messages`. Not 100% happy with this name, but the `add_to_justified_messages` method was doing way more than that.

Reply:
also, put the common features of the method in the parent. Binary doesn't even override this method now. Blockchain calls super and does some extra work.